New lint: `iter_overeager_cloned`

pmnoxx opened the pull request:
Closes #8202 

changelog: New lint: [`iter_overeager_cloned`]

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @llogiq (or someone else) soon.

Please see [the contribution instructions](https://github.com/rust-lang/rust-clippy/blob/master/CONTRIBUTING.md) for more information.

llogiq commented on clippy_lints/src/methods/cloned_last.rs:
```diff
@@ -0,0 +1,36 @@
+            expr.span,
+            msg,
+            "try this",
+            format!("{}.last().cloned()", iter_snippet),
```
If the `.last()` is `unwrapped` or `?`d, we could even suggest to `.clone()` that. Just a suggestion for extending the lint, this won't stop us from merging.

pmnoxx replied:
@llogiq Actually, you are right. The result of `last()` is an `Option`, so we should be using `close()` instead.

I updated the PR.

pmnoxx replied:
@llogiq That's a great suggestion. 

I see:  for `Option<>` /  `Result<>`.
`recv.cloned().unwrap()`
could be
`recv.unwrap().clone`.

I see that `cloned` should be moved furthermost the chain as long as possible, not just for `last`.
For example here:
```rust
 vec.iter().cloned().filter(...).skip(5).take(10).next().last().unwrap()
```
That could be just written as
```rust
let val = vec.cloned().filter(...).skip(5).take(10).next().last().unwrap().clone()
```

This may even allow us to drop the `clone()` entirely, in some cases, when clone() is not needed.

It may be a good idea to propagate clone from before, to after
- `last`
- `take`
- `skip`
- `next`.
- `unwrap` 
- `filter` // has lambda, so requires some more checks
- `find`.// has lambda,so requires some more checks
- ... probably some others as well

The question is, should each be a separate lint, or should they be combined together?

llogiq replied:
I think your example is a bit off, but otherwise full ack.

I'd be OK with multiple lints if there are useful distinctions to make. E.g. if one part of the check can have false positives, it would be a good idea to put that into a different `nursery` lint. Or if one part could reasonably be `style` and another `perf`. Otherwise, putting all the checks into one lint seems natural. One thing I that comes to mind is that we already have the `needless_clone` lint, so we could add cases where the `cloned` can be removed outright to that, making the rest `early_iter_cloned` or some such.

Again, I'm also OK if you restrict the scope of this PR and do a followup later. Have fun!

llogiq reviewed (approved):
This looks good to me. I have an idea how we could make that lint even better. If you want to do that within this PR, go ahead, otherwise let me know and I'll r+.

llogiq commented on clippy_lints/src/methods/cloned_last.rs:
```diff
@@ -0,0 +1,32 @@
+
+/// lint use of `cloned().next()` for `Iterators`
+pub(super) fn check<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx hir::Expr<'_>, recv: &'tcx hir::Expr<'_>) {
+    // lint if caller of `.filter().next()` is an Iterator
```
This comment looks out of place here.

pmnoxx replied:
@llogiq Thanks, I fixed the comment.

llogiq commented on clippy_lints/src/methods/cloned_last.rs:
```diff
@@ -0,0 +1,31 @@
+
+use super::CLONED_LAST;
+
+/// lint use of `cloned().next()` for `Iterators`
```
Shouldn't that be ` cloned().last()`?

On the other hand, you could extend the lint to `needless_cloned_iter` and lint `.last()`, `.next()` and `.count()` (the latter simply removing `cloned()`...

llogiq commented:
I'd be OK with merging this with the `.cloned().next()` replaced by `.cloned().last()` in the two comments, or if you want you can tackle making the lint more powerful. Ping me if you need any help!

llogiq commented on tests/ui/cloned_last.rs:
```diff
@@ -7,4 +7,16 @@ fn main() {
         .cloned()
         .last();
+
+    #[rustfmt::skip]
```
Why skip rustfmt?

pmnoxx replied:
I copied it from another example. It's not really needed, so I removed it

pmnoxx commented:
> I'd be OK with merging this with the `.cloned().next()` replaced by `.cloned().last()` in the two comments, or if you want you can tackle making the lint more powerful. Ping me if you need any help!

I added implementation for
- skip
- take
- count
- last
- next
- nth


TODO
- update documentation

Questions:
- what name should be used for new lint? `needless_cloned` ?
- should it be applied to `clippy::all` or `clippy::nursery`
- maybe this should only be applies to types ,which are bigger than 8 bytes, as it's more efficient to deference types 8 bytes and shorter

I think that support for `.cloned().next().unwrap()` should be in a separate PR. 

 Fixing `option_receiver.cloned().unwrap()` or `result_receiver.cloned().unwrap()`, shouldn't be hard to to either, I'll take a look at it later.

llogiq commented:
My name suggestion as per comment above is: Add the cases where the `cloned` call can be removed outright to `needless_clone`, the others could be `needless_iter_cloned` or `early_iter_cloned` or `overeager_iter_cloned`. For `last` and `skip`, the latter seems to be most appropriate.

llogiq commented:
Oh, and I would like to run some tests before deciding whether to put it in `style`, `perf` or `nursery`. If we find false positives or lots of noise, I'd suggest the latter. Otherwise I'd be OK with either of the former.

llogiq commented on tests/ui/cloned_last.rs:
```diff
@@ -0,0 +1,36 @@
+#![warn(clippy::all)]
+
+fn main() {
+    let _: Option<String> = vec!["1".to_string(), "2".to_string(), "3".to_string()]
```
How about `["1", "2", "3"].into_iter().cloned().collect::<Vec<_>>()` to avoid the `to_string()`s?

Also perhaps put it in a binding and clone as needed.

pmnoxx replied:
```rust
143 |                     let x = ["1", "2", "3"].into_iter().cloned().collect::<Vec<_>>();
    |                                                         ^^^^^^ the trait `Clone` is not implemented for `str`
```

pmnoxx replied:
I could write it like this:
```
let _ = ["1", "2", "3"].into_iter().map(ToString::to_string).collect::<Vec<_>>()
```

pmnoxx replied:
Although, that would deny the purpose of having this test.

llogiq commented on tests/ui/cloned_last.rs:
```diff
@@ -0,0 +1,36 @@
+        .iter()
+        .cloned()
+        .nth(2);
+}
```
I'd like to see some tests against false positives if possible: Methods after `.cloned()` which are OK, and if you do anything that would require more complex tests (e.g. `.map(_)`), there should be tests for those checks, too.

pmnoxx replied:
@llogiq I added some examples, where doing automatic replacement would be bad, because the signature of methods would change.

I added a list of TODOs as well.

llogiq commented on clippy_lints/src/methods/mod.rs:
```diff
@@ -107,6 +108,29 @@ declare_clippy_lint! {
 
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for usage of `_.cloned().last()`.
```
The documentation should also reflect the complete lint.

pmnoxx replied:
@llogiq I updated the documentation.

llogiq commented on clippy_lints/src/methods/mod.rs:
```diff
@@ -107,6 +108,29 @@ declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for usage of `_.cloned().last()`.
+    ///
+    /// ### Why is this bad?
```
Not only performance, but also style: For `.cloned().next()`, the call will very likely be inlined anyway, but as code changes, keeping a consistent style is also a good thing.

I'd also add a `### Known Problems` section, because the lint is partially behavior-changing, because `.cloned().last()` will clone all elements unlike `.last().cloned()`. This cloning could be observed as a side effect, which would be removed by this lint. Given the contract of `Clone` and the likely perf benefit, I find this acceptable, but we should still document it.

llogiq commented on clippy_lints/src/methods/mod.rs:
````diff
@@ -107,6 +108,29 @@ declare_clippy_lint! {
+    /// ```
+    #[clippy::version = "1.59.0"]
+    pub CLONED_LAST,
+    complexity,
````
I think this should be `perf` or `nursery`, depending on the number of additional warnings in lintcheck vs. master. If it's too noisy, I'd prefer the latter, otherwise I'd look into the added warnings and see if the benefit outweighs the added effort to fix the lint.

pmnoxx replied:
@llogiq It looks to me that maybe the lint should be split. Sometimes it may not make a sense to do it, if it's just `.cloned().next()`, though it depend on whats after.

Other times it may be worth it, because you may avoid cloning the whole collection.

pmnoxx replied:
I was planning to write a more complicated check, which can cover more cases, but I think this PR is a good start. There are a few things that I don't know how to use yet.

llogiq replied:
I don't think that would warrant a split: Just because the code has `.next()` now doesn't mean that it won't get changed to `.nth(_)` later. Also even when only changing `.cloned().next()`, to `.next().cloned()`, this will reduce compile time if ever so slightly, so that's a perf benefit anyway.

pmnoxx commented:
> My name suggestion as per comment above is: Add the cases where the `cloned` call can be removed outright to `needless_clone`, the others could be `needless_iter_cloned` or `early_iter_cloned` or `overeager_iter_cloned`. For `last` and `skip`, the latter seems to be most appropriate.

I agree, as a use of clippy I think like to see it divided maybe into
- `needless_cloned` - `count`, `len`, `is_empty`   - `cloned` is remoed
- `overeager_iter_cloned` - `skip`, `take`, `last`, `nth` - affects performance; `cloned` is used less often` - has side effect
-  `early_iter_cloned` - `next` - clone can be done later, perhaps this will allow further optimizations later on; - moves clone further into the chain, may allow further performance optimizations. 

- `filter` - can be added later to  `overeager_iter_cloned`
- `map`. `flat_map` - not sure, sometimes may be `map` / `flat_map` doesn't need cloned version of the item.

llogiq commented:
I'm ok with the `needless_clone` / `overeager_iter_cloned` split.

I would avoid adding `needless_cloned`, because a) that's *very* similar to `needless_clone` which we already have and b) also conceptually similar enough we may as well just extend that lint.

I don't think the `early_iter_cloned` warrants it's own lint. As I explained above, code changes, and doing it now has no drawbacks and benefits compile time even if only slightly.

Note that we'll have to extend the `needless_clone` documentation to include the cases if we go that route.

pmnoxx commented on clippy_lints/src/methods/iter_overeager_cloned.rs:
```diff
@@ -0,0 +1,78 @@
+        implements_trait(cx, cx.typeck_results().expr_ty(recv), id, &[])
+    });
+    if recv_impls_iterator {
+        let (lint, msg) = match name {
+            "count" => (
+                REDUNDANT_CLONE,
+                format!(
+                    "called `cloned().{}()` on an `Iterator`. It may be more efficient to call\
+                    `.{}()` instead",
+                    name, name
```
@llogiq I'm not sure if this is the preferred way to return `REDUNDANT_CLONE` lint.

If not I'll change the code of `redundant_clone.rs`.

llogiq replied:
Why would it not be? Issuing the lint here keeps the code simple. Perhaps add a comment to the `redundant_clone` code that links to the implementation.

pmnoxx replied:
> Why would it not be? Issuing the lint here keeps the code simple. Perhaps add a comment to the `redundant_clone` code that links to the implementation.

@llogiq This depends on `clippy` design; Whenever or or all `check` always run or not. I suspect that's the case, but I didn't read that part of the code to make sure.

pmnoxx replied:
And also on how checks are used. I don't know whenever, there is a rule that says "all lints should be generated from file with the same name". It was the case for some lints, but I didn't check if that's true. Though, I see you don't see a problem, so I guess it's fine.

pmnoxx commented:
@llogiq I think some more changes are needed.

--) Cloning small values, references of references is actually beneficial. 

Consider following 
```rust
/// cloned() should not be removed
let _ = [0u8, 0u8].iter().cloned()....
/// cloned() should not be removed
let _ = Vec::<&&String>.iter().cloned()....
```
Cloning `&u8` is a good idea, this will improve performance, as there is no need to do dereferencing.
Same thing can be said about `&&String`. 

I think we should apply the same rules as `clippy_lints/src/pass_by_ref_or_value.rs` uses.

If the type after cloning is `<=8 bytes`, then we should keep the `cloned`, otherwise postpone it.
For example, changing `&&String` to `&String`, keep the size the same, but removes one extra deference needed.

llogiq commented:
That should be `.copied()`, not `.cloned()`, but we already have that lint. Regardless, not linting if the element type is of `size <= size_of::<usize>()` makes sense.

pmnoxx commented:
> That should be `.copied()`, not `.cloned()`, but we already have that lint. Regardless, not linting if the element type is of `size <= size_of::<usize>()` makes sense.

I added a test, and I see that this would be applied automatically:
```
diff --git i/tests/ui/iter_overeager_cloned.fixed w/tests/ui/iter_overeager_cloned.fixed
index 01b64ff7a..7295ad871 100644
--- i/tests/ui/iter_overeager_cloned.fixed
+++ w/tests/ui/iter_overeager_cloned.fixed
@@ -39,4 +39,7 @@ fn main() {
 
     // Not implemented yet
     let _ = vec.iter().cloned().any(|x| x.len() == 1);
+
+    // Should probably stay as it is.
+    let _ = [0, 1, 2, 3, 4].iter().take(10).cloned();
 }
diff --git i/tests/ui/iter_overeager_cloned.rs w/tests/ui/iter_overeager_cloned.rs
index fede4c733..dd04e33a4 100644
--- i/tests/ui/iter_overeager_cloned.rs
+++ w/tests/ui/iter_overeager_cloned.rs
@@ -41,4 +41,7 @@ fn main() {
 
     // Not implemented yet
     let _ = vec.iter().cloned().any(|x| x.len() == 1);
+
+    // Should probably stay as it is.
+    let _ = [0, 1, 2, 3, 4].iter().cloned().take(10);
 }
diff --git i/tests/ui/iter_overeager_cloned.stderr w/tests/ui/iter_overeager_cloned.stderr
index 4a58687f4..343614e75 100644
--- i/tests/ui/iter_overeager_cloned.stderr
+++ w/tests/ui/iter_overeager_cloned.stderr
@@ -54,5 +54,11 @@ LL ~     let _ = [Some(Some("str".to_string())), Some(Some("str".to_string()))]
 LL ~         .iter().flatten().cloned();
    |
 
-error: aborting due to 7 previous errors
+error: called `cloned().take(...)` on an `Iterator`. It may be more efficient to call`.take(...).cloned()` instead
+  --> $DIR/iter_overeager_cloned.rs:46:13
+   |
+LL |     let _ = [0, 1, 2, 3, 4].iter().cloned().take(10);
+   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try this: `[0, 1, 2, 3, 4].iter().take(10).cloned()`
+
+error: aborting due to 8 previous errors
```

llogiq commented on clippy_lints/src/methods/mod.rs:
```diff
@@ -107,6 +108,42 @@ declare_clippy_lint! {
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for usage of `_.cloned().<func>()` where call to `.cloned()` can be postponed,
+    /// or often removed entirely.
```
Don't we issue `REDUNDANT_CLONE` when the `.cloned()` can be removed?

pmnoxx replied:
@llogiq Ok, fixed.

llogiq commented:
Cool stuff! My remaining nits are with the documentation, otherwise this looks great.

llogiq commented on clippy_lints/src/methods/iter_overeager_cloned.rs:
```diff
@@ -0,0 +1,78 @@
+    name: &str,
+    map_arg: &[hir::Expr<'_>],
+) {
+    // lint if caller of `.cloned().last()` is an Iterator
```
```suggestion
    // lint if callee is an Iterator
```

llogiq commented on clippy_lints/src/methods/iter_overeager_cloned.rs:
```diff
@@ -0,0 +1,78 @@
+use super::ITER_OVEREAGER_CLONED;
+use crate::redundant_clone::REDUNDANT_CLONE;
+
+/// lint use of `cloned().last()` for `Iterators`
```
```suggestion
/// lint overeager use of `cloned()` for `Iterator`s
```

llogiq reviewed (commented):
A few more nits to fix. Do you need help with the size check?

pmnoxx commented:
> A few more nits to fix. Do you need help with the size check?

I'm still thinking whenever the check is needed. For example, for

- `count` - 

> pass_by_ref_or_value

I'm not entirely convinced whenever the check is needed. For example.

```rust
[0u8, 1u8, 2u8, 3u8].iter().cloned().count() /// cloned() should be removed
 [0, 1, 2, 3, 4].last().cloned().last();          /// it's better to not deference values, and move cloned() last
 [0, 1, 2, 3, 4].last().cloned().nth(40);          /// same here
 [0, 1, 2, 3, 4].last().cloned().skip(40);          /// same here
 [0, 1, 2, 3, 4].last().cloned().next();     // probably it's better not to print - not sure
 [0, 1, 2, 3, 4].last().cloned().take(50);     // probably it's better not to print - not sure
 [0, 1, 2, 3, 4].last().cloned().flatten();     // probably it's better not to print - not sure
```

@llogiq I'm not sure how to get  the type from the iterator. I'll have to spend some time and figure out a better way to navigate through code. Type hinting doesn't seem to work correctly with Rust plugin for Intellij, which makes things harder.

My understanding is that the receiver is some generic iterator, that implements `Iterator<Item = SomeItem>`. I'm not sure yet how to do get it.

llogiq commented:
I just had a discussion with someone from libs. There's a Rust PR [#90209](https://github.com/rust-lang/rust/pull/90209) that aims to remove more clones through specialization. Apparently the libs team thinks that as there never was any guarantee of current behavior, they're free to change it. Which is fine by me. That somewhat reduces the need for this lint, but doesn't render it moot, because as we all know specialization is a fickle beast and may sometimes fail to apply. So we should still add this lint, but perhaps note the possibility of not seeing a perf win in the docs.

bors commented:
:umbrella: The latest upstream changes (presumably #8198) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

llogiq commented:
That PR was closed, so we're left with the clippy lint and some docs on `Iterator::cloned` I just wrote. Hopefully this will suffice to have people avoid most needless clones on iterating.

pmnoxx commented:
> That PR was closed, so we're left with the clippy lint and some docs on `Iterator::cloned` I just wrote. Hopefully this will suffice to have people avoid most needless clones on iterating.

Ok, just let me know what you think is best. Whenever we should ships this PR as it is. Or improve it, or close this PR.

llogiq commented:
Re-reading this, I think we can look into refining the lint later. For now let's just merge it.

@bors r+

bors commented:
:pushpin: Commit e209ff8a008ec609844b09059458806d9bb40fee has been approved by `llogiq`

<!-- @bors r=llogiq e209ff8a008ec609844b09059458806d9bb40fee -->
<!-- homu: {"type":"Approved","sha":"e209ff8a008ec609844b09059458806d9bb40fee","approver":"llogiq"} -->

bors commented:
:lock: Merge conflict

This pull request and the master branch diverged in a way that cannot be automatically merged. Please rebase on top of the latest master branch, and let the reviewer approve again.

<details><summary>How do I rebase?</summary>

Assuming `self` is your fork and `upstream` is this repository, you can resolve the conflict following these steps:

1. `git checkout piotr-next-lint` *(switch to your branch)*
2. `git fetch upstream master` *(retrieve the latest master)*
3. `git rebase upstream/master -p` *(rebase on top of it)*
4. Follow the on-screen instruction to resolve conflicts (check `git status` if you got lost).
5. `git push self piotr-next-lint --force-with-lease` *(update this PR)*

You may also read [*Git Rebasing to Resolve Conflicts* by Drew Blessing](http://blessing.io/git/git-rebase/open-source/2015/08/23/git-rebasing-to-resolve-conflicts.html) for a short tutorial.

Please avoid the ["**Resolve conflicts**" button](https://help.github.com/articles/resolving-a-merge-conflict-on-github/) on GitHub. It uses `git merge` instead of `git rebase` which makes the PR commit history more difficult to read.

Sometimes step 4 will complete without asking for resolution. This is usually due to difference between how `Cargo.lock` conflict is handled during merge and rebase. This is normal, and you should still perform step 5 to update this PR.

</details>

<details><summary>Error message</summary>

```text
Auto-merging clippy_lints/src/methods/mod.rs
CONFLICT (content): Merge conflict in clippy_lints/src/methods/mod.rs
Auto-merging clippy_lints/src/lib.register_perf.rs
Auto-merging clippy_lints/src/lib.register_lints.rs
Auto-merging clippy_lints/src/lib.register_all.rs
Auto-merging CHANGELOG.md
Automatic merge failed; fix conflicts and then commit the result.

```

</details>

llogiq commented:
@bors r-

llogiq commented:
I overlooked that you need to rebase the PR on current master. Shouldn't be too hard.

pmnoxx commented:
> I overlooked that you need to rebase the PR on current master. Shouldn't be too hard.

I rebased the PR.

llogiq commented:
@bors r+

bors commented:
:pushpin: Commit 36396c6b972714bda6c25ab14a7a954136b8654f has been approved by `llogiq`

<!-- @bors r=llogiq 36396c6b972714bda6c25ab14a7a954136b8654f -->
<!-- homu: {"type":"Approved","sha":"36396c6b972714bda6c25ab14a7a954136b8654f","approver":"llogiq"} -->

bors commented:
:hourglass: Testing commit 36396c6b972714bda6c25ab14a7a954136b8654f with merge 93cad4a94d296aa6f0870f9e3e8d89635ac060e6...
<!-- homu: {"type":"BuildStarted","head_sha":"36396c6b972714bda6c25ab14a7a954136b8654f","merge_sha":"93cad4a94d296aa6f0870f9e3e8d89635ac060e6"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/runs/4833944485?check_suite_focus=true), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/runs/4833938070?check_suite_focus=true), [checks-action_test](https://github.com/rust-lang/rust-clippy/runs/4834011455?check_suite_focus=true)
Approved by: llogiq
Pushing 93cad4a94d296aa6f0870f9e3e8d89635ac060e6 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"llogiq","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/runs/4834011455?check_suite_focus=true","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/runs/4833944485?check_suite_focus=true","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/runs/4833938070?check_suite_focus=true"},"merge_sha":"93cad4a94d296aa6f0870f9e3e8d89635ac060e6"} -->
